src/data/zmod/basic.lean

```diff
@@ -3,526 +3,706 @@ Copyright (c) 2018 Chris Hughes. All rights reserved.
-@zmod.val_cast_of_lt ⟨p, hp.pos⟩ _ h
+@[simp] lemma inv_coe_unit {n : ℕ} (u : units (zmod n)) :
+  (u : zmod n)⁻¹ = (u⁻¹ : units (zmod n)) :=
+begin
+  have := congr_arg (coe : ℕ → zmod n) (val_coe_unit_coprime u),
+  rw [← mul_inv_eq_gcd, nat.cast_one] at this,
+  let u' : units (zmod n) := ⟨u, (u : zmod n)⁻¹, this, by rwa mul_comm⟩,
+  have h : u = u', { apply units.ext, refl },
+  rw h,
+  refl
+end
 
-@[simp] lemma cast_mod_nat (a : ℕ) : ((a % p : ℕ) : zmodp p hp) = a :=
-@zmod.cast_mod_nat ⟨p, hp.pos⟩ _
+lemma mul_inv_of_unit {n : ℕ} (a : zmod n) (h : is_unit a) :
+  a * a⁻¹ = 1 :=
+begin
+  rcases h with ⟨u, rfl⟩,
+  rw [inv_coe_unit, u.mul_inv],
+end
 
-@[simp] lemma cast_val (a : zmodp p hp) : (a.val : zmodp p hp) = a :=
-@zmod.cast_val ⟨p, hp.pos⟩ _
+lemma inv_mul_of_unit {n : ℕ} (a : zmod n) (h : is_unit a) :
+  a⁻¹ * a = 1 :=
+by rw [mul_comm, mul_inv_of_unit a h]
+
+/-- Equivalence between the units of `zmod n` and
+the subtype of terms `x : zmod n` for which `x.val` is comprime to `n` -/
+def units_equiv_coprime {n : ℕ} [fact (0 < n)] :
+  units (zmod n) ≃ {x : zmod n // nat.coprime x.val n} :=
+{ to_fun := λ x, ⟨x, val_coe_unit_coprime x⟩,
+  inv_fun := λ x, unit_of_coprime x.1.val x.2,
+  left_inv := λ ⟨_, _, _, _⟩, units.ext (cast_val _),
+  right_inv := λ ⟨_, _⟩, by simp }
 
-@[simp] lemma cast_mod_int (a : ℤ) : ((a % p : ℤ) : zmodp p hp) = a :=
-@zmod.cast_mod_int ⟨p, hp.pos⟩ _
+section totient
+open_locale nat
 
-lemma val_cast_int (a : ℤ) : (a : zmodp p hp).val = (a % p).nat_abs :=
-@zmod.val_cast_int ⟨p, hp.pos⟩ _
+@[simp] lemma card_units_eq_totient (n : ℕ) [fact (0 < n)] :
+  fintype.card (units (zmod n)) = φ n :=
+calc fintype.card (units (zmod n)) = fintype.card {x : zmod n // x.val.coprime n} :
+  fintype.card_congr zmod.units_equiv_coprime
+... = φ n :
+begin
+  apply finset.card_congr (λ (a : {x : zmod n // x.val.coprime n}) _, a.1.val),
+  { intro a, simp [a.1.val_lt, a.2.symm] {contextual := tt}, },
+  { intros _ _ _ _ h, rw subtype.ext, apply val_injective, exact h, },
+  { intros b hb,
+    rw [finset.mem_filter, finset.mem_range] at hb,
+    refine ⟨⟨b, _⟩, finset.mem_univ _, _⟩,
+    { let u := unit_of_coprime b hb.2.symm,
+      exact val_coe_unit_coprime u },
+    { show zmod.val (b : zmod n) = b,
+      rw [val_cast_nat, nat.mod_eq_of_lt hb.1], } }
+end
 
-lemma coe_val_cast_int  (a : ℤ) : ((a : zmodp p hp).val : ℤ) = a % (p : ℕ) :=
-@zmod.coe_val_cast_int ⟨p, hp.pos⟩ _
+end totient
 
-lemma eq_iff_modeq_nat {a b : ℕ} : (a : zmodp p hp) = b ↔ a ≡ b [MOD p] :=
-@zmod.eq_iff_modeq_nat ⟨p, hp.pos⟩ _ _
+instance subsingleton_units : subsingleton (units (zmod 2)) :=
+⟨λ x y, begin
+  cases x with x xi,
+  cases y with y yi,
+  revert x y xi yi,
+  exact dec_trivial
+end⟩
 
-lemma eq_iff_modeq_int {a b : ℤ} : (a : zmodp p hp) = b ↔ a ≡ b [ZMOD p] :=
-@zmod.eq_iff_modeq_int ⟨p, hp.pos⟩ _ _
+lemma le_div_two_iff_lt_neg (n : ℕ) [hn : fact ((n : ℕ) % 2 = 1)]
```

Comment:
I am a little bit uneasy with a complicated typeclass assumption such as this one. Why not state it as a regular assumption? In the applications, is it really provided by typeclass inference?

Reply:
Yeah, I made the decision that `odd/even` was still ok. But for example I kept `p % 4 = 3` as a regular assumption.
I'm fine with reverting this one. @ChrisHughes24 What do you think?

Reply:
I would only use these instance brackets, if it's actually used to synthesize an instance in the statement.

Reply:
Also, we have a definition `even` now, so you could write `\not even`

Reply:
Hmm, I don't know how well `even` and `\not even` play together with `n % 2`. If they play well together, I guess that might be worth another refactor PR.